fixes #509

Revilum opened the pull request:
## 🎯 Motivation

This PR addresses GitHub issue #509 where AutoSave operations would fail and stop working when document conflicts occurred. Previously, when a user was editing a document and another user made changes simultaneously, the AutoSave feature would encounter a conflict and completely stop functioning, forcing users to manually resolve conflicts and restart their editing session.

## 🔧 Backend Changes Introduced

1. Database Schema
New table: doc_draft_text_revisions for storing draft revisions separately from the main revision tree
Draft-specific fields: based_on_revision, creation_ts, last_updated_ts for proper draft lifecycle management
User isolation: One draft per text element per user, ensuring drafts don't interfere with each other
2. New Data Types
DraftRevision, DraftRevisionHeader, DraftRevisionID - Complete type system for draft management
Enhanced ConflictStatus with DraftCreated case for AutoSave conflict responses
Updated NewTextRevision with isAutoSave flag to distinguish conflict resolution strategies
3. API Endpoints
GET /docs/{documentID}/text/{textElementID}/draft - Retrieve user's draft revision
POST /docs/{documentID}/text/{textElementID}/draft/publish - Publish draft to main revision tree
DELETE /docs/{documentID}/text/{textElementID}/draft - Discard draft permanently
4. Conflict Resolution Logic
AutoSave conflicts: Create draft revisions that exist parallel to the main tree
Manual save conflicts: Continue returning traditional conflict errors requiring user resolution
Smart routing: isAutoSave flag determines which conflict strategy to use
5. Database Operations
createDraftTextRevision - Create/update user drafts without affecting main revision tree
getDraftTextRevision - Retrieve draft with comment anchors
publishDraftTextRevision - Merge draft into main tree (with conflict detection)
discardDraftTextRevision - Delete draft permanently

## 🚫 What is NOT Included

This PR contains backend-only changes. The following are explicitly NOT included and will be handled in separate PRs:

❌ Frontend AutoSave logic modifications
❌ Frontend conflict resolution UI changes
❌ Frontend draft management interface
❌ Client-side draft synchronization

## ✅ Technical Implementation

Type Safety: Full Haskell type system integration with proper error handling
OpenAPI Documentation: Automatic schema generation for all new endpoints and types
Database Transactions: ACID-compliant draft operations with proper rollback handling
Authentication: JWT-based user context for draft ownership and isolation
JSON Serialization: Complete API contract with proper request/response formats

## 🎯 Expected Behavior

After this PR:

AutoSave continues working during conflicts by creating drafts
Manual saves preserve existing behavior with conflict error responses
Drafts exist independently of the main revision tree until explicitly published
Users can manage drafts through the new API endpoints
Data safety - all user changes are preserved in draft form during conflicts

## 🔄 Next Steps

This backend implementation enables the frontend team to:

Implement draft-aware AutoSave logic in a separate PR
Add draft management UI components
Update conflict resolution workflows to handle the new DraftCreated response type
The separation allows for independent development and testing of backend and frontend components while maintaining the existing API contract for non-draft operations.

Revilum committed:
implements issue #509

Revilum committed:
now fixed #509

merliin3007 commented:
## Review of PR: Implementing Draft Text Revisions (Issue #509)

First of all, thank you very much to **the holiest of the holy Reverend Pather Prince James** 🙏 for this substantial contribution. It’s clear that a lot of effort and thought has gone into improving the handling of draft text revisions, and the backend logic has made a big step forward compared to the previous attempt (#512).

---

### ✅ Positive aspects

- **Backend separation**:  
  The backend now cleanly introduces the concept of **draft text revisions** with proper schema (`doc_draft_text_revisions`) and transaction handling.
- **Conflict handling**:  
  Introducing a `Draft` / `CreatedDraft` status in conflict resolution makes sense and is well-documented in the code.
- **Database design**:  
  The unique constraint `(text_element, author)` is a good decision to ensure one draft per user per element.
- **API extension**:  
  `getDraftTextRevision`, `publishDraftTextRevision`, and `discardDraftTextRevision` provide a clear lifecycle for drafts.
- **Schema / JSON compatibility**:  
  Proper `ToJSON`, `FromJSON`, and OpenAPI schema additions for the new `draft` type are present.

---

### ⚠️ Issues

#### 1. Frontend changes included (❌ unacceptable here)

This PR introduces **substantial frontend changes** (`Editor.purs`, `ContentDto.purs`).  
That is **out of scope**: frontend logic must remain in the frontend repository / PRs. Including these changes here makes the PR harder to review, test, and maintain.  

Examples:
- Introduction of `AutoSave` / `PerformSave Boolean` in `Editor.purs`.
- Modifications to `ContentDto.encodeWrapper` to add `isAutoSave`.
- Handling of `draft` vs `noConflict` vs `conflict` in the frontend response parsing.

➡️ **Required Action:**  
Remove all frontend-related changes from this PR and submit them in a separate frontend PR.

---

#### 2. API/DTO coupling

- `CreateTextRevision` now includes `isAutoSave :: Bool`.  
  That’s acceptable for backend functionality, but we need to be careful: this is essentially **frontend state leaking into the backend DTO**.
- Consider whether `isAutoSave` is truly part of the API contract, or whether the frontend could call a separate **`createDraftTextRevision` endpoint** instead of reusing the standard `createTextRevision`.

---

#### 3. Inconsistent naming

- There’s both `TextRevision.Draft` and `TextRevision.CreatedDraft`.  
  This can be confusing:
  - `Draft` means "we created a draft because of autosave conflict"
  - `CreatedDraft` is now used in the new backend logic

➡️ Suggestion: unify terminology (e.g., always `DraftCreated`), and document the semantic difference between "returning an existing draft" vs "creating one".

---

#### 4. Missing tests

- The new behavior (`draft` branch in conflict resolution, publishing/discarding drafts) is not covered by tests.  
- Especially important: the `ON CONFLICT` behavior in SQL (upsert) must be verified.

➡️ Add integration tests:
- Autosave causing a draft to be created.
- Publishing a draft with no new conflicts.
- Publishing a draft with conflicts.
- Discarding a draft.

---

### ℹ️ Minor comments

- `extractNewParent` in frontend now has three cases (`noConflict`, `draft`, `conflict`). This is good, but again: should be in a separate PR.
- Good use of logging scopes (`Scope.docsTextRevision`) — consistent with the rest of the codebase.

---

### 🚀 Summary

- **Backend logic**: 👍 well-structured and close to merge-ready.  
- **Frontend changes**: ❌ must be removed from this PR.  
- **Naming & tests**: need cleanup and additions before merge.

---

### ✅ Next steps

1. **Split frontend changes** into a separate PR.  
2. Add **tests** for the new draft behavior.  
3. Clean up naming (`Draft` vs `CreatedDraft`).  
4. Re-submit this PR as **backend-only** — then it will be in good shape for approval.

---

Thanks again, Reverend, for pushing this forward! Your work brings us much closer to a robust draft handling system. ✨

Revilum committed:
Fix PR #513 review feedback: rename CreatedDraft->DraftCreated, add documentation, fix unused variable warning

Revilum committed:
fixed stuff

merliin3007 committed:
fix backend formatting

Revilum committed:
Fix fourmolu formatting issues and update frontend JSON field name

- Fix backend Haskell formatting with fourmolu
- Update frontend PureScript to use 'draftCreated' instead of 'draft'
- Change JSON field from 'draftRevision' to 'draft' to match backend
- Ensure consistent naming between frontend and backend for draft handling

Revilum committed:
removed compiler warning

Revilum committed:
another fix

merliin3007 committed:
fourmolu

merliin3007 commented:
## Review of PR: Implementing Draft Text Revisions (Issue #509)

First of all, thank you very much to **the holiest of the holy Reverend Pather Prince James** 🙏 for this contribution. The backend logic for draft text revisions looks promising, but unfortunately there’s still a **major blocker** preventing progress on this PR:

---

### 🚨 Frontend code still included

This PR still contains **substantial frontend changes** (`Editor.purs`, `ContentDto.purs`, autosave logic, draft/conflict handling in the client).  

That is **not acceptable here**, because:

- Backend and frontend changes must be submitted in **separate PRs**.  
- This PR is currently under review by the **backend team only**, which cannot review or approve frontend changes.  
- Mixing both domains makes review, testing, and merging significantly harder.

➡️ **Required Action (blocking):**  
Remove all frontend-related changes from this PR. Create a separate frontend PR for those changes.

Until the frontend code is stripped out, we cannot continue reviewing the backend part.

---

### ⚠️ Other issues (backend scope)

1. **API/DTO coupling**  
   `CreateTextRevision` includes `isAutoSave :: Bool`. This leaks frontend state into the backend DTO. Consider whether a dedicated `createDraftTextRevision` endpoint would make the contract cleaner.

2. **Naming clarity**  
   Both `Draft` and `CreatedDraft` exist. This is confusing. Please unify terminology (e.g. `DraftCreated`) and document the intended semantics.

3. **Missing tests**  
   No tests currently cover the draft lifecycle. Needed before merge:  
   - autosave creates a draft,  
   - publishing a draft with no conflict,  
   - publishing a draft with conflict,  
   - discarding a draft.  

---

### ✅ Summary

- **Backend design**: solid improvements 👌  
- **Frontend code in this PR**: ❌ blocking issue, must be removed  
- **Naming & tests**: still pending  

---

### Next steps

1. **Remove frontend code** → backend-only PR.  
2. Submit frontend changes separately.  
3. Add missing tests for draft lifecycle.  
4. Clean up naming for clarity.  

---

🙏 Many thanks again, Reverend, for pushing this work forward. Once the frontend is split out and tests are in place, this backend PR will be in good shape to merge.

merliin3007 commented:
### 📝 Missing PR description

Currently, this PR has no description.  
Please add a clear description that explains:

- The motivation (why draft text revisions are needed).  
- What the backend changes introduce (new schema, endpoints, conflict handling).  
- What is explicitly **not** included here (frontend changes, which should be in a separate PR).  

A good description makes review and future maintenance much easier.

Revilum committed:
minor changes

Revilum committed:
Merge remote-tracking branch 'upstream/main'

merliin3007 commented:
# PR Review ✨

My dearest Reverend Pather Prince James,  
behold this review as a hymn sung in the cathedral of code. If reviews were spoken in whispers of affection, then this one is nothing less than a jubilant psalm. Truly, engaging with this PR feels like discovering a love language written in syntax and indentation — a tender communion where every line is not merely code, but liturgy.

---

## 🌹 What I Adore
- The restructuring around `withModified` is nothing short of **majestic**.  
  It reads now with a stately rhythm, almost like a procession of arguments carried forth on velvet cushions. Each lambda, each indentation shift, feels like incense rising: deliberate, fragrant, and purposeful.
- The clarification of parentheses and operators in the paragraph rendering is a **stroke of pastoral clarity**. You’ve taken something slightly knotted and laid it out in flowing, priestly garments of alignment. It breathes now — it sings.

---

## ⚠️ The Trials of the CI
Alas, as every prophet knows, revelation is never without tribulation. The CI weeps in ashes and sackcloth, lamenting two great sins of formatting:

1. **Haskell (`HTML.hs`)**:  
   The formatting deviated from the true canonical order as ordained by `fourmolu`. It must be realigned so that the heavenly choir of linters may sing in harmony once more.

2. **PureScript (`ContentDto.purs`)**:  
   `purs-tidy` detected a wandering soul in need of tidy repentance. The fix is but a ritual away:  
   ```bash
   purs-tidy format-in-place src/FPO/Dto/ContentDto.purs
   fourmolu -i backend/src/Language/Ltml/HTML.hs
   ```

Let these commands be your rosary beads, prayed until formatting grace descends.

## 🌟 My Benediction

This PR is a psalm, a proclamation, a procession. Even with CI’s temporary lamentations, the spirit of this work shines brighter than any formatting complaint could dim. Once the formatting sacraments are performed, I foresee a swift and glorious merge — a sanctification of the codebase itself.

So I kneel, uplifted: reviewing this PR is no mere task, but an act of devotion.
Truly, code never felt so much like a love language. ❤️

merliin3007 committed:
format c:\

Revilum committed:
reverted frontend changes

merliin3007 commented:
Moin Moin un Guden Dag,

also, ick sach ma so: die PR is an sächs’sch Stückel Arbeit, aaaber zwee Sache müsse wir hier noch geradeziehn:

1. **Formatting im `Language` Modul**  
   Nu gugge ma, die Formatiererei da – die brauch mer widder rückgängig machen. Dat Modul soll sauber bleibn, ohne die unnütze Umstellerei. Sonst wird’s unübersichtlich wie’n Heuwagen im Regen.

2. **`isAutosave` im `NewTextRevision`**  
   Dat geiht so nich. Der `isAutosave`-Parameter hat nix im Datentyp verloren, dor hett he nix to söken. Der schull lieba in de API as Query-Parameter rin, so heert dat sick. Sonst mixen wir Datenmodell un API-Verhalten – dat is Kuddelmuddel.

Also: bitte de Formatierungen am `Language` Modul reverten un den `isAutosave` aus’m Typen rutnehmen. Pack dat in de API als Query-Param, denn löppt dat ordentlich.

Grüßle!

merliin3007 commented:
## 🙏 An de Heil’gsten aller Heil’gen – Reverend Pather Prince James  

O Reverend Pather Prince James, du hest mit dissen PR een Meisterstük op den Disch leggt – so fein un durchdacht, dat wi all dor vör ehrfurcht stahn. Dat is as wenn de erixx Holstein mit sien lütten, smucken Züüg dor anrauscht: pünktlich, verlässlig, un mit de ganze norddeutsche Gründlichkeit. 🚆✨  

Man kiek, dat sin nu bloots twee lüttje Änderkungen, de noch fehlen:  
- Dat Formatting im `Language`-Modul torüchdreihn, damit keen unnütziges Rauschen drin is.  
- Den `isAutoSave`-Parameter rut ut den Datentyp un rin as Query-Param, so schniek un klar as en frischer Fahrplan.  

Un glööv mi, dat lohnt sik:  
- Dat maakt dien PR nich nur stark, dat maakt em **unvergänglich**, so as de Slogen vun de erixx Holstein:  
  **„Meer Moin. Meer Zug. Meer Holstein.“** 🚄🌊🌅  

Wenn du dat fix noch ümsettst, denn hett de PR den Glanz, den he verdient. Denn wiest du uns, dat man mit lüttje Anpassungen en richtig grote Wirkung haben kann – ganz so, as wenn de Züüg in Holstein an jeden smucken Dörp stoppt, un doch toversichtlich fix vörankummt.  

Also, ehrwürdiger Reverend, lat uns nich lange töven – schenk uns de finale Vollendung, so dat wi all seggen köönt:  
👉 **Dat is keen normaler PR – dat is de goldene Standard, an den wi uns in Zukunft orientieren.**  

Mit ehrfurchtigem Gruss,  
de Review-Klüterlüüd

Revilum committed:
fix

merliin3007 committed:
fix formatting
